feat: redirect agents to backing controller

kian99 opened the pull request:
## Description
This change is in support of model migration. During model migration, agents connect to the new controller to validate they will work after migration has completed. JIMM should not be the address for Juju agents, they should point to the backing controller but the migration flow needs to go through JIMM for it to modify things like the model owner. To handle this, when an agent is trying to login to JIMM we return a redirect to the backing controller.

Both integration and unit tests have been added and JIMM's errors package has also been modified to support all the error fields that are used in Juju's RPC error handling/

## Engineering checklist
<!-- *Check only items that apply* -->

- [ ] Documentation updated
- [x] Covered by unit tests
- [x] Covered by integration tests

kian99 committed:
feat: redirect agents to backing controller

This change is in support of model migration. During model migration, agents connect to the new controller to validate they will work after migration has completed.
JIMM should not be the address for Juju agents, they should point to the backing controller but the migration flow needs to go through JIMM for it to modify things like the model owner.
To handle this, when an agent is trying to login to JIMM we return a redirect to the backing controller.

alesstimec commented on internal/jujuapi/websocket_test.go:
```diff
@@ -223,6 +224,44 @@ func (s *apiProxySuite) TestSessionTokenLoginProvider(c *gc.C) {
+	}
+	_, err = api.Open(&info, dialOpts)
+	c.Assert(err, gc.NotNil)
+	_, ok := errors.Cause(err).(*api.RedirectError)
```
could we assert that the redirect data matches what we expect?

alesstimec commented on internal/rpcproxy/rpcproxy.go:
```diff
@@ -35,6 +36,22 @@ const (
+// ControllerDetails holds information about the controller
+// that is being proxied to.
+type ControllerDetails struct {
+	Servers [][]jujuparams.HostPort
```
Addresses?

alesstimec commented on internal/rpcproxy/rpcproxy.go:
```diff
@@ -35,6 +36,22 @@ const (
+// when we need to redirect a client to that controller
+// instead of proxying the request. This is the case during
+// model migration when receiving requests from agents.
+type RedirectInfo interface {
```
suggestion: RedirectInfoStore? RedirectInfoProvider?

kian99 replied:
I'll go with `RedirectInfoGetter`

alesstimec reviewed (approved):
LGTM with a few suggestions

kian99 committed:
chore: rename servers to addresses

kian99 committed:
fix: add info field to wellknown handler tests

The JWKS endpoints return a JSON object that is just mapping our internal error type.
We shouldn't do this, there is no specification that requires our JWKS endpoint to return a JSON formatted error and we especially shouldn't just marshal our internal error type to JSON.
This will need to be addressed separately in a follow-up.

kian99 committed:
chore: fix rpcproxy tests

SimoneDutto commented on internal/rpcproxy/rpcproxylogin_test.go:
```diff
@@ -467,3 +528,34 @@ func (m *mockTokenGenerator) GetUser() names.UserTag {
+	}, nil
+}
+
+func expectedRedirectInfo() map[string]interface{} {
```
how is this working if now in the error we return `Addresses` and not `servers`?

kian99 replied:
`servers` is what Juju expects, `Addresses` is just the field for our internal type.
More specifically,
The model proxy talks to the DB and gets a `ControllerDetails` struct with the `Addresses` field populated.
It then populates a `jujuparams.RedirectErrorInfo` which serializes to an object with a field called `servers`.

kian99 replied:
I tested changing `servers` to `addresses` and it does fail as expected,
```
              + 		"addresses": []any{
              + 			[]any{
              + 				map[string]any{
              + 					"port":  float64(17070),
              + 					"scope": string(""),
              + 					"type":  string("ipv4"),
              + 					"value": string("controller-1"),
              + 				},
              + 			},
              + 		},
                		"ca-cert": string("test-ca-cert"),
              - 		"servers": []any{
              - 			[]any{
              - 				map[string]any{
              - 					"port":  float64(17070),
              - 					"scope": string(""),
              - 					"type":  string("ipv4"),
              - 					"value": string("controller-1"),
              - 				},
              - 			},
              - 		},
                	},
```

SimoneDutto reviewed (commented):
it looks good, just a question about the tests

SimoneDutto reviewed (approved):
lgtm
